modules/currency-l1/src/main/scala/org/tessellation/currency/l1/domain/snapshot/programs/CurrencySnapshotProcessor.scala

```diff
@@ -0,0 +1,169 @@
+                  case (NonEmptyList(snapshot, nextSnapshots), agg) =>
+                    checkAlignment(snapshot, bs, lcss).flatMap {
+                      case _: Ignore =>
+                        Applicative[F].pure(none[SuccessT].asRight[LeftT])
+                      case alignment =>
+                        processAlignment(snapshot, alignment, bs, ts, lcss, as).as {
+                          val updatedAgg = agg :+ (alignment, snapshot)
+
+                          NonEmptyList.fromList(nextSnapshots) match {
+                            case Some(next) =>
+                              (next, updatedAgg).asLeft[RightT]
+                            case None =>
+                              NonEmptyList
+                                .fromList(updatedAgg)
+                                .asRight[LeftT]
+                          }
+                        }
+                    }
+                }
+            }.flatMap {
+              case Some(alignmentsAndSnapshots) =>
+                alignmentsAndSnapshots.traverse {
+                  case (alignment, hashedSnapshot) =>
+                    processAlignment(
+                      hashedSnapshot,
+                      alignment,
+                      blockStorage,
+                      transactionStorage,
+                      lastCurrencySnapshotStorage,
+                      addressStorage
+                    )
+                }.flatMap { results =>
+                  setGlobalSnapshot
+                    .map(BatchResult(_, results))
+                }
+              case None => Applicative[F].pure(SnapshotIgnored(globalSnapshotReference))
+            }
+
+          case Some(Validated.Invalid(_)) =>
+            Slf4jLogger
+              .getLogger[F]
+              .warn(s"Not all currency snapshots are signed correctly! Ignoring global snapshot: $globalSnapshotReference")
+              .as(SnapshotIgnored(globalSnapshotReference))
+
+          case None =>
+            setGlobalSnapshot
+        }
+
+      private def prepareIntermediateStorages(
+        addressStorage: AddressStorage[F],
+        blockStorage: BlockStorage[F, CurrencyBlock],
+        lastCurrencySnapshotStorage: LastSnapshotStorage[F, CurrencySnapshot],
+        transactionStorage: TransactionStorage[F, CurrencyTransaction]
+      ): F[
+        (
+          AddressStorage[F],
+          BlockStorage[F, CurrencyBlock],
+          LastSnapshotStorage[F, CurrencySnapshot],
+          TransactionStorage[F, CurrencyTransaction]
+        )
+      ] = {
+        val bs = blockStorage.getState().flatMap(BlockStorage.make[F, CurrencyBlock](_))
+        val lcss = lastCurrencySnapshotStorage.get.flatMap(LastSnapshotStorage.make(_))
+        val as = addressStorage.getState().flatMap(AddressStorage.make(_))
+        val ts =
+          transactionStorage.getState().flatMap { case (lastTxs, waitingTxs) => TransactionStorage.make(lastTxs, waitingTxs) }
```

Comment:
Could you tell me more about these intermediate storages? I mean I am wondering why these are created inside the processor.